In [1]:
from movielens_analysis import Movies, Tags, Links, Ratings

# initialize all classes
%timeit Movies("../datasets/movies.csv")
movies = Movies("../datasets/movies.csv")
%timeit Tags("../datasets/tags.csv")
tags = Tags("../datasets/tags.csv")
%timeit Links("../datasets/links.csv",50)
links = Links("../datasets/links.csv",50)
%timeit Ratings("../datasets/ratings.csv", "../datasets/movies.csv")
rating = Ratings("../datasets/ratings.csv", "../datasets/movies.csv")


22.5 ms ± 2.31 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
2.14 ms ± 347 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
230 μs ± 19.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
13.4 ms ± 446 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# MOVIES

In [2]:
%timeit len(movies.movies)
print(f'Total in database {len(movies.movies)} movies')

67.4 ns ± 2.65 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
Total in database 9742 movies


Let's see which years were of key importance.

In [3]:
%timeit movies.dist_by_release()
years=movies.dist_by_release()
print(years)
print(f"Let's find the year when the most films were released:")
%timeit for year, count in list(movies.dist_by_release().items())[:5]: pass
for year, count in list(years.items())[:5]:
    print(f"{year}: {count} films")
max_year=next(iter(years))
print(f"Why is {max_year} year have the most films been released?")

17.5 ms ± 714 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
{2002: 311, 2006: 295, 2001: 294, 2007: 284, 2000: 283, 2009: 282, 2003: 279, 2004: 279, 2014: 278, 1996: 276, 2015: 274, 2005: 273, 2008: 269, 1999: 263, 1997: 260, 1995: 259, 1998: 258, 2011: 254, 2010: 247, 2013: 239, 1994: 237, 2012: 233, 2016: 218, 1993: 198, 1992: 167, 1988: 165, 1987: 153, 1990: 147, 1991: 147, 2017: 147, 1989: 142, 1986: 139, 1985: 126, 1984: 101, 1981: 92, 1980: 89, 1982: 87, 1983: 83, 1979: 69, 1977: 63, 1973: 59, 1978: 59, 1965: 47, 1971: 47, 1974: 45, 1976: 44, 1964: 43, 1967: 42, 1968: 42, 1975: 42, 1966: 42, 2018: 41, 1962: 40, 1972: 39, 1963: 39, 1959: 37, 1960: 37, 1955: 36, 1969: 35, 1961: 34, 1970: 33, 1957: 33, 1958: 31, 1953: 30, 1956: 30, 1940: 25, 1949: 25, 1954: 23, 1942: 23, 1939: 23, 1946: 23, 1951: 22, 1950: 21, 1947: 20, 1948: 20, 1941: 18, 1936: 18, 1945: 17, 1937: 16, 1952: 16, 1944: 16, 1938: 15, 1931: 14, 1935: 13, 1933: 12, 1934: 11, 1943: 10, 1932: 9, 1927: 7, 1930: 

In [4]:
print(f"Let's take a look at 10 films that were released in 2002.")
%timeit movies.get_movies_by_year(2002)[:10]
movies_in_most_year=movies.get_movies_by_year(2002)[:10]
print(*movies_in_most_year, sep="\n")

Let's take a look at 10 films that were released in 2002.
13.2 ms ± 362 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
"11'09""01 - September 11 (2002)"
"Adventures of Pluto Nash, The (2002)"
"Anarchist Cookbook, The (2002)"
"Bang, Bang, You're Dead (2002)"
"Banger Sisters, The (2002)"
"Bourne Identity, The (2002)"
"Cat Returns, The (Neko no ongaeshi) (2002)"
"Cat's Meow, The (2002)"
"Century of the Self, The (2002)"
"Crime of Father Amaro, The (Crimen del padre Amaro, El) (2002)"


### And now, let's analyze the genres of films

In [5]:
%timeit movies.dist_by_genres()
genres_dist = movies.dist_by_genres()
%timeit for genre, count in movies.dist_by_genres().items(): pass
print("Let's look at the distribution of films by genre:")
for genre, count in genres_dist.items():
    print(f"{genre}: {count} films")

6.51 ms ± 286 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.62 ms ± 419 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Let's look at the distribution of films by genre:
Drama: 4361 films
Comedy: 3756 films
Thriller: 1894 films
Action: 1828 films
Romance: 1596 films
Adventure: 1263 films
Crime: 1199 films
Sci-Fi: 980 films
Horror: 978 films
Fantasy: 779 films
Children: 664 films
Animation: 611 films
Mystery: 573 films
Documentary: 440 films
War: 382 films
Musical: 334 films
Western: 167 films
IMAX: 158 films
Film-Noir: 87 films


The most popular genre is drama. As many as 4,361 films were released that included this genre!

In [6]:
%timeit movies.most_genres(10)
multi_genre_movies = movies.most_genres(10)
print(f"Let's look at the top 10 films in which the largest number of genres were indicated:")
for title, count in multi_genre_movies.items():
    print(f"{title}: {count} geners")

5.05 ms ± 513 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Let's look at the top 10 films in which the largest number of genres were indicated:
Rubber (2010): 10 geners
Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989): 8 geners
Mulan (1998): 7 geners
Who Framed Roger Rabbit? (1988): 7 geners
Osmosis Jones (2001): 7 geners
Interstate 60 (2002): 7 geners
Robots (2005): 7 geners
Pulse (2006): 7 geners
Aqua Teen Hunger Force Colon Movie Film for Theaters (2007): 7 geners
Enchanted (2007): 7 geners


This is the movie "Rubber" from 2010! It is not surprising that it was included in the list.
![..\datasets\kinopoisk.jpg](..\datasets\kinopoisk.jpg)
This movie has even more genres in kinopoisk!

In [7]:
print(f"Let's see how many genres per movie are listed on average.")
%timeit sum(len([g for g in m['genres'] if g != '(no genres listed)']) for m in movies.movies) / len(movies.movies)
avg_geners_per_movie=sum(len([g for g in m['genres'] if g != '(no genres listed)']) for m in movies.movies) / len(movies.movies)
print(f"Avarage amount  genres per movie: {avg_geners_per_movie:.2f}")

Let's see how many genres per movie are listed on average.
8.03 ms ± 385 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Avarage amount  genres per movie: 15.90


# Tags

In [8]:
%timeit top_tags_result_timed = tags.most_words(10)
top_tags = tags.most_words(10)
print(f"Top 10 tags with the most words:")
for tag, word_count in top_tags.items():
    print(f"{tag} : {word_count} words")

378 μs ± 30.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Top 10 tags with the most words:
Something for everyone in this one... saw it without and plan on seeing it with kids! : 16 words
the catholic church is the most corrupt organization in history : 10 words
Oscar (Best Music - Original Score) : 6 words
based on a true story : 5 words
Everything you want is here : 5 words
lord of the rings : 4 words
seen more than once : 4 words
based on a play : 4 words
jay and silent bob : 4 words
political right versus left : 4 words


Wow, some user wrote a tag with as many as 10 words!

### Now let's look at the longest tag relative to characters, not words.

In [9]:
%timeit top_tags_result_timed = tags.longest(10)
top_tags = tags.longest(10)
print(f"Top 10 tags with the most characters:")
for tag, word_count in top_tags.items():
    print(f"{tag} : {word_count} слов")

254 μs ± 11.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Top 10 tags with the most characters:
Something for everyone in this one... saw it without and plan on seeing it with kids! : 85 слов
the catholic church is the most corrupt organization in history : 63 слов
audience intelligence underestimated : 36 слов
Oscar (Best Music - Original Score) : 35 слов
assassin-in-training (scene) : 28 слов
Oscar (Best Cinematography) : 27 слов
political right versus left : 27 слов
Everything you want is here : 27 слов
representation of children : 26 слов
artificial intelligence : 23 слов


Note that several of them match, but not all of them, we need to find their intersection.

In [10]:
%timeit top_tags_result_timed = tags.most_words_and_longest(10)
top_tags = tags.most_words_and_longest(10)
print(f"Intersection of the top 10 tags with the most characters and words:")
print(*top_tags, sep='\n')

473 μs ± 30.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Intersection of the top 10 tags with the most characters and words:
Everything you want is here
Oscar (Best Music - Original Score)
Something for everyone in this one... saw it without and plan on seeing it with kids!
the catholic church is the most corrupt organization in history


### And now it's time to find the most popular tags.

In [11]:
%timeit top_popular_tags_result_timed = tags.most_popular(10)
top_popular_tags = tags.most_popular(10)
print(f"Top 10 most popular tags:")
for tag, count in top_popular_tags.items():
    print(f"{tag} : {count} counts")

169 μs ± 2.56 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Top 10 most popular tags:
funny : 15 counts
sci-fi : 14 counts
twist ending : 12 counts
dark comedy : 12 counts
atmospheric : 10 counts
superhero : 10 counts
comedy : 10 counts
action : 10 counts
suspense : 10 counts
Leonardo DiCaprio : 9 counts


Wow, the most popular tag is "funny". We need to find films that have this tag.

In [12]:
%timeit tags.movie_by_tag('funny')
%timeit {movie['movieId']: movie['title'] for movie in movies.movies}
movie_dict = {movie['movieId']: movie['title'] for movie in movies.movies}
movie_id=tags.movie_by_tag('funny')
%timeit for movie_ids in movie_id: pass
print("Movies with the 'funny' tag:")
for movie_ids in movie_id:
    title = movie_dict.get(movie_ids, "Unknown movie")
    print(f"{title}")

62.7 μs ± 1.01 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
1.06 ms ± 32.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
235 ns ± 10.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Movies with the 'funny' tag:
"Croods, The (2013)"
Anchorman 2: The Legend Continues (2013)
Big Hero 6 (2014)
The Interview (2014)
The DUFF (2015)
Jumanji: Welcome to the Jungle (2017)
Game Night (2018)
Home Alone 2: Lost in New York (1992)
Toy Story 2 (1999)
Clueless (1995)
Step Brothers (2008)
"Brothers Bloom, The (2008)"
Zombieland (2009)
Friends with Benefits (2011)
Django Unchained (2012)


There are many popular films among these films, so we can say that people like funny films.

### How about looking at the word search tags?

In [13]:
search_word = "Oscar"
%timeit matching_tags_result_timed = tags.tags_with(search_word)
matching_tags = tags.tags_with(search_word)
%timeit for tag in matching_tags: pass
print(f"Tags containing the word '{search_word}':")
for tag in matching_tags:
    print(f"{tag}")

67.4 μs ± 4.16 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
94.3 ns ± 3.76 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
Tags containing the word 'Oscar':
Oscar (Best Actress)
Oscar (Best Cinematography)
Oscar (Best Music - Original Score)


### And now it's time to find the most popular tags.

In [14]:
%timeit top_popular_tags_result_timed = tags.most_popular(10)
top_popular_tags = tags.most_popular(10)
print(f"Top 10 most popular tags:")
for tag, count in top_popular_tags.items():
    print(f"{tag} : {count} counts")

166 μs ± 2.1 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Top 10 most popular tags:
funny : 15 counts
sci-fi : 14 counts
twist ending : 12 counts
dark comedy : 12 counts
atmospheric : 10 counts
superhero : 10 counts
comedy : 10 counts
action : 10 counts
suspense : 10 counts
Leonardo DiCaprio : 9 counts


# LINKS

Class instanciation: I modified constructor a bit, so by default it uses first 1000 ids as task authors suggets, but u can use second arg, which I strongly recommend to do, cause fetchin 1000 takes like 25 mins. So lets stick with 50

In [15]:
%timeit links.get_ids_dict()
links_ids=links.get_ids_dict()
%timeit for num,id in links_ids.items(): pass
print("Let's look at the ID of the films that we will be loking")
for num,id in links_ids.items():
    print(f"{num} : {id}")

81.2 ns ± 3.2 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
1.52 μs ± 123 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Let's look at the ID of the films that we will be loking
1 : 0114709
2 : 0113497
3 : 0113228
4 : 0114885
5 : 0113041
6 : 0113277
7 : 0114319
8 : 0112302
9 : 0114576
10 : 0113189
11 : 0112346
12 : 0112896
13 : 0112453
14 : 0113987
15 : 0112760
16 : 0112641
17 : 0114388
18 : 0113101
19 : 0112281
20 : 0113845
21 : 0113161
22 : 0112722
23 : 0112401
24 : 0114168
25 : 0113627
26 : 0114057
27 : 0114011
28 : 0114117
29 : 0112682
30 : 0115012
31 : 0112792
32 : 0114746
34 : 0112431
36 : 0112818
38 : 0113442
39 : 0112697
40 : 0112749
41 : 0114279
42 : 0112819
43 : 0114272
44 : 0113855
45 : 0114681
46 : 0113347
47 : 0114369
48 : 0114148
49 : 0114916
50 : 0114814
52 : 0113819
53 : 0110299
54 : 0112499


instance.get_imdb(list_of_movies, list_of_fields) returns required parsed data, but first call(from any place in code) fetches all pages for our id pool and parses em for fields demanded for our methods to minimalize net usage and delay, savin in self right after. But here we have to wait mentioned fetchin time, for 50 ids its aprox 1 min

In [16]:
%timeit links.get_imdb([],[])
links.get_imdb([],[])

The slowest run took 29.00 times longer than the fastest. This could mean that an intermediate result is being cached.
3.77 μs ± 6.76 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)


[]

Time passed and now we can get any info we need truly fast, lets start with top 7 longest films, simply call longest(10)

In [17]:
%timeit links.longest(10)
longest_movies=links.longest(10)
%timeit for name, minutes in longest_movies.items(): pass
print(f'Top-10 of longest films for our piece of data is:')
for name, minutes in longest_movies.items():
    print(f"{name} : {minutes} minutes")

65.7 μs ± 3.9 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
491 ns ± 66.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Top-10 of longest films for our piece of data is:
Nixon : 192 minutes
Casino : 178 minutes
Heat : 170 minutes
Sense and Sensibility : 136 minutes
Assassins : 133 minutes
GoldenEye : 130 minutes
12 Monkeys : 129 minutes
Se7en : 127 minutes
Sabrina : 127 minutes
Cutthroat Island : 124 minutes


Now let's find the most expensive movies.

In [18]:
%timeit links.most_expensive(10)
most_expensive_movies=links.most_expensive(10)
%timeit for name, minutes in most_expensive_movies.items(): pass
print(f'Top-10 of most expensive films for our piece of data is:')
for name, cost in most_expensive_movies.items():
    print(f"{name} : {cost} $")

48.7 μs ± 3.5 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
395 ns ± 17.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Top-10 of most expensive films for our piece of data is:
Cutthroat Island : 98000000.0 $
Money Train : 68000000.0 $
Jumanji : 65000000.0 $
The American President : 62000000.0 $
GoldenEye : 60000000.0 $
Heat : 60000000.0 $
Sabrina : 58000000.0 $
Pocahontas : 55000000.0 $
Casino : 52000000.0 $
Assassins : 50000000.0 $


Ok, what about most profitable movies?

In [19]:
%timeit links.most_profitable(10)
profitable_movies=links.most_profitable(10)
%timeit for name, minutes in profitable_movies.items(): pass
print(f'Top-10 of most expensive films for our piece of data is:')
for name, cost in profitable_movies.items():
    print(f"{name} : {cost} $")

59.7 μs ± 3.01 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
403 ns ± 13.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Top-10 of most expensive films for our piece of data is:
Toy Story : 364436586.0 $
Se7en : 295983304.0 $
GoldenEye : 292194034.0 $
Pocahontas : 291079773.0 $
Babe : 224134910.0 $
Jumanji : 197821940.0 $
Ace Ventura: When Nature Calls : 182385533.0 $
Dangerous Minds : 156519401.0 $
12 Monkeys : 139839459.0 $
Heat : 127436818.0 $


Cool, but are there any intersections between them?

In [20]:
print(*set(most_expensive_movies) & set(profitable_movies), sep='\n')

GoldenEye
Heat
Jumanji
Pocahontas


Now we know that only 3 expensive films have become the most profitable

What about top directors?

In [21]:
%timeit links.top_directors(10)
top_directors=links.top_directors(10)
%timeit for name, minutes in top_directors.items(): pass
print(f'Top-10 of top directors for our piece of data is:')
for director, num in top_directors.items():
    print(f"{director} : {num} ")

52.3 μs ± 6.04 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
400 ns ± 13.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Top-10 of top directors for our piece of data is:
Holly Goldberg Sloan : 1 
Gianni Amelio : 1 
Woody Allen : 1 
Bryan Singer : 1 
Patricia Rozema : 1 
Mike Gabriel : 1 
David Fincher : 1 
Jocelyn Moorhouse : 1 
Gus Van Sant : 1 
Paul W.S. Anderson : 1 


lets see top cost per minute for movies

In [22]:
%timeit links.top_cost_per_minute(10)
top_cost_per_minute=links.top_cost_per_minute(10)
%timeit for name, minutes in top_cost_per_minute.items(): pass
print(f'Top-10 of top directors for our piece of data is:')
for name, cost in top_cost_per_minute.items():
    print(f"{name} : {cost} $")

81.4 μs ± 3.01 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
389 ns ± 9.65 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Top-10 of top directors for our piece of data is:
Cutthroat Island : 790322.58 $
Pocahontas : 679012.35 $
Jumanji : 625000.0 $
Money Train : 618181.82 $
The American President : 543859.65 $
GoldenEye : 461538.46 $
Sabrina : 456692.91 $
Assassins : 375939.85 $
Toy Story : 370370.37 $
Heat : 352941.18 $


Now, lets see ratings of the fisrt movie of our parse of data.

In [23]:
print(*links.get_imdb_rating([1]).values())

8.3/10


## Ratings class

Join ratings.csv and movies.csv for getting data. getting info about users and films.

All ratings are contained in the file `ratings.csv`. Each line of this file after the header row represents one rating of one movie by one user, and has the following format:
    userId,movieId,rating,timestamp

Movie information is contained in the file `movies.csv`. Each line of this file after the header row represents one movie, and has the following format:

    movieId,title,genres

import Ratings class from ratings.py and initialixing data sets movies and users_ratings.
Ratings.Movies Inner Class

__init__(self, parent)

Initializes the Movies analysis class, using the parent Ratings instance.

In [24]:
%timeit Ratings("../datasets/ratings.csv", "../datasets/movies.csv")
rating = Ratings("../datasets/ratings.csv", "../datasets/movies.csv")
%timeit  rating.Movies(rating)
movies = rating.Movies(rating)
%timeit rating.Users(rating)
users_ratings = rating.Users(rating)

14.5 ms ± 1.09 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
267 ns ± 17.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
381 ns ± 17.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


dist_by_year(self)
Returns a dictionary where keys are years (extracted from rating timestamps) and values are counts of ratings per year.
Sorted by year ascending.

In [25]:
%timeit (movies.dist_by_year())
print(movies.dist_by_year())

585 μs ± 29.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
{1996: 358, 1999: 82, 2000: 296, 2001: 70, 2005: 121, 2006: 4, 2007: 1, 2011: 39, 2015: 29}


dist_by_rating(self)
Returns a dictionary where keys are rating values and values are counts of each rating.
Sorted by rating ascending.

In [26]:
%timeit movies.dist_by_rating()
print(movies.dist_by_rating())

156 μs ± 6.75 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
{0.5: 24, 1.0: 39, 1.5: 11, 2.0: 57, 2.5: 7, 3.0: 253, 3.5: 17, 4.0: 292, 4.5: 33, 5.0: 267}


top_by_num_of_ratings(self, n)
Returns the top-n movies by the number of ratings.
Keys: movie titles
Values: number of ratings
Sorted by count descending.

In [27]:
%timeit movies.top_by_num_of_ratings(10)
print(movies.top_by_num_of_ratings(10))

240 μs ± 14 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
{'Usual Suspects, The (1995)': 4, 'Pulp Fiction (1994)': 4, 'Fugitive, The (1993)': 4, "Schindler's List (1993)": 4, 'Batman (1989)': 4, 'Silence of the Lambs, The (1991)': 4, 'Fargo (1996)': 4, 'Aladdin (1992)': 4, 'Beauty and the Beast (1991)': 4, 'Toy Story (1995)': 3}


top_by_ratings(self, n, metric='average')
Returns the top-n movies by the average or median of their ratings.
metric can be 'average' or 'median'
Keys: movie titles
Values: average or median rating (rounded to 2 decimals)
Sorted by metric descending.

In [28]:
%timeit movies.top_by_ratings(1, 'median')
print(movies.top_by_ratings(1, 'median'))

1.22 ms ± 28.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
{'Bottle Rocket (1996)': 5.0}


top_controversial(self, n)
Returns the top-n movies by the variance of their ratings.
Keys: movie titles
Values: variance (rounded to 2 decimals)
Sorted by variance descending.

In [29]:
%timeit movies.top_controversial(1)
print(movies.top_controversial(1))

1.62 ms ± 148 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
{'Bambi (1942)': 5.06}


most_active_user_by_coverage(self)
Returns a tuple (userId, percent) for the user who rated the largest percentage of movies in the sample.
percent is rounded to 2 decimals.

In [30]:
%timeit movies.most_active_user_by_coverage()
print(movies.most_active_user_by_coverage())

241 μs ± 12.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
(6, 39.15)


In [31]:
%timeit movies.percent_of_max_ratings_per_movie(1)
print(movies.percent_of_max_ratings_per_movie(1))

892 μs ± 52.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
{'Bottle Rocket (1996)': 100.0}


users_distribution(self)
Returns a dictionary where keys are user IDs and values are the number of ratings each user has given.

In [32]:
%timeit users_ratings.users_distribution()
print(users_ratings.users_distribution())

101 μs ± 2.84 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
{1: 232, 2: 29, 3: 39, 4: 216, 5: 44, 6: 314, 7: 126}


users_rating_distribution(self, metric='average')
Returns a dictionary where keys are user IDs and values are the average or median of their ratings (rounded to 2 decimals).
metric can be 'average' or 'median'.

In [33]:
%timeit users_ratings.users_distribution()
print(users_ratings.users_distribution())

103 μs ± 3.84 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
{1: 232, 2: 29, 3: 39, 4: 216, 5: 44, 6: 314, 7: 126}


top_n_users_by_variance(self, n)
Returns the top-n users by the variance of their ratings.
Keys: user IDs
Values: variance (rounded to 2 decimals)
Sorted by variance descending.

In [34]:
%timeit users_ratings.users_rating_distribution()
print(users_ratings.users_rating_distribution())

141 μs ± 2.24 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
{1: 4.37, 2: 3.95, 3: 2.44, 4: 3.56, 5: 3.64, 6: 3.49, 7: 3.35}
